WIP: Binary Outcome DLC

nkohen opened the pull request:
This PR builds on previous Schnorr work to construct a DLC with a binary outcome between two participants.

(Was previously #683)

nkohen committed:
Rebased secp256k1 jni commits on top of jonas/schnorrsig (#657)

An attempt to add schnorr_sign and schnorr_verify to the jni. All tests pass!

nkohen committed:
Defined BinaryOutcomeDLC and attempted to construct CETs with a new MultiSignatureWithTimeoutScriptPubKey

Used P2WSH in DLC outputs

skeleton of DLC execution

nkohen committed:
Schnorr Sign with Specified Nonce (#665)

* Added schnorrSignWithNonce to jni, nonces can now be specifed (tests pass)

* Added NativeSecp256k1.schnorrPublicNonce which computes the R value for a given k

* Responded to code review

torkelrogstad commented on core/src/main/scala/org/bitcoins/core/crypto/Schnorr.scala:
```diff
@@ -0,0 +1,71 @@
+  def sign(
+      msg: ByteVector,
+      privateKey: ECPrivateKey): SchnorrDigitalSignature = {
+    if (msg.length != 32) {
```
`require` instead? You're basically hand rolling it here

torkelrogstad commented on core/src/main/scala/org/bitcoins/core/crypto/Schnorr.scala:
```diff
@@ -0,0 +1,71 @@
+  def signWithNonce(
+      msg: ByteVector,
+      privateKey: ECPrivateKey,
+      nonce: ECPrivateKey): SchnorrDigitalSignature = {
```
`nonce` here is a priv key because it needs to be 32 bytes, right? Not sure how I feel about using `ECPrivateKey`, as types are also very important to convey meaning to params etc

torkelrogstad commented on core/src/main/scala/org/bitcoins/core/crypto/Schnorr.scala:
```diff
@@ -0,0 +1,71 @@
+      msg: ByteVector,
+      privateKey: ECPrivateKey,
+      nonce: ECPrivateKey): SchnorrDigitalSignature = {
+    if (msg.length != 32) {
```
Same thing with `require` from abvoe

torkelrogstad commented on core/src/main/scala/org/bitcoins/core/crypto/Schnorr.scala:
```diff
@@ -0,0 +1,71 @@
+      msg: ByteVector,
+      sig: SchnorrDigitalSignature,
+      publicKey: ECPublicKey): Boolean = {
+    if (msg.length != 32) {
```
`require`

torkelrogstad commented on core/src/main/scala/org/bitcoins/core/crypto/Schnorr.scala:
```diff
@@ -0,0 +1,71 @@
+      msg: ByteVector,
+      r: ECPublicKey,
+      publicKey: ECPublicKey): ECPublicKey = {
+    if (msg.length != 32) {
```
`require`

torkelrogstad commented on core/src/main/scala/org/bitcoins/core/crypto/SchnorrDigitalSignature.scala:
```diff
@@ -0,0 +1,59 @@
+import org.bitcoins.core.util.{BitcoinSLogger, BitcoinSUtil, Factory}
+import scodec.bits.ByteVector
+
+sealed abstract class SchnorrDigitalSignature extends BitcoinSLogger {
```
IMO there's no real advantage to using the sealed trait/factory object/impl case class pattern here, might as well define `SchnorrDigitialSignature` as a case class (which should lead to fewer allocations as well)

torkelrogstad commented on core/src/main/scala/org/bitcoins/core/crypto/SchnorrDigitalSignature.scala:
```diff
@@ -0,0 +1,59 @@
+      extends SchnorrDigitalSignature
+
+  override def fromBytes(bytes: ByteVector): SchnorrDigitalSignature = {
+    if (bytes.length != 64) {
```
`require`

torkelrogstad commented on core/src/main/scala/org/bitcoins/core/crypto/SchnorrDigitalSignature.scala:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  def apply(rx: ByteVector, s: ByteVector): SchnorrDigitalSignature = {
+    if (rx.length != 32) {
```
`require`. Also, this is covered by the `require` statements in the abstract class

torkelrogstad commented on core/src/main/scala/org/bitcoins/core/crypto/SchnorrDigitalSignature.scala:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  def apply(r: ECPublicKey, s: ByteVector): SchnorrDigitalSignature = {
+    if (s.length != 32) {
```
this length check is covered in the abstract class

torkelrogstad commented on core/src/main/scala/org/bitcoins/core/protocol/script/ScriptPubKey.scala:
```diff
@@ -255,6 +255,87 @@ object MultiSignatureScriptPubKey
+    extends ScriptFactory[MultiSignatureWithTimeoutScriptPubKey] {
+  import org.bitcoins.core.script.control._
+
+  private case class MultiSignatureWithTimeoutScriptPubKeyImpl(
```
Same thing as above. Do we need the abstract class/case class indirection?

torkelrogstad commented on dlc/README.md:
```diff
@@ -0,0 +1 @@
+#TODO
```
The other modules don't really have meaningful READMEs, FYI. Probably makes more sense to move this to the website docs, adding a section on DLCs

torkelrogstad committed:
Publish schnorr branch with updated names (#703)

torkelrogstad committed:
New attempt at schnorr branch Travis config (#706)

torkelrogstad committed:
2019 08 21 travis schnorr take 2 (#707)

* New attempt at schnorr branch Travis config

* New attempt at schnorr branch Travis config - take 2

torkelrogstad committed:
New attempt at schnorr branch Travis config - take 3 (#709)

nkohen committed:
Secp256k1 Support for DLCs (#682)

* Added computeSchnorrPubKey to the jni and added a property test which passses!

Wrote more extensive tests and they pass

Created Schnorr object as an interface for scala code, as well as introducing a SchnorrDigitalSignature type

* Added test for private key leakage on two messages being signed

* Created SchnorrNonce class and responded to code review

* Tidied things up

* Removed factory pattern from SchnorrDigitalSignature

* Added comment narration of Schnorr tests

* Responded to code review

* nits

Made snapshot publishing work on schnorr-dlc

nkohen commented:
Replaced by #871
